tests/test_main.py

```diff
@@ -1,6 +1,324 @@
-# TODO: add tests
-from app.main import hello_world
+import inspect
+import io
+from contextlib import redirect_stdout
 
+import pytest
 
-def test_hello_world():
-    assert hello_world() == "Hello, world!"
+from app.main import Player, Elf, Dwarf, ElfRanger, Druid, DwarfWarrior, DwarfBlacksmith, calculate_team_total_rating, \
+    elves_concert, feast_of_the_dwarves
+
+
+@pytest.mark.parametrize(
+    "class_,methods",
+    [
+        (Player, ["get_rating", "player_info"]),
+        (Elf, ["get_rating", "player_info", "play_elf_song"]),
+        (ElfRanger, ["get_rating", "player_info", "play_elf_song"]),
+        (Druid, ["get_rating", "player_info", "play_elf_song"]),
+        (Dwarf, ["get_rating", "player_info", "eat_favourite_dish"]),
+        (DwarfWarrior, ["get_rating", "player_info", "eat_favourite_dish"]),
+        (DwarfBlacksmith, ["get_rating", "player_info", "eat_favourite_dish"]),
+    ],
+)
+def test_classes_should_have_corresponding_methods(class_, methods):
+    for method in methods:
+        assert (
+                hasattr(class_, method) is True
+        ), f"Class '{class_.__name__}' should have method {method}"
+
+
+@pytest.mark.parametrize(
+    "class_",
+    [
+        Player, Elf, Dwarf,
+    ]
+)
+def test_classes_should_be_abstract(class_):
+    assert inspect.isabstract(class_) is True, (
+        f"Class '{class_.__name__}' should be abstract"
+    )
+
+
+@pytest.mark.parametrize(
+    "class_",
+    [
+        ElfRanger, Druid, DwarfWarrior, DwarfBlacksmith,
+    ]
+)
+def test_classes_should_not_be_abstract(class_):
+    assert inspect.isabstract(class_) is False, (
+        f"Class '{class_.__name__}' shouldn't be abstract"
+    )
+
```

Comment:
Also here you could add checks about correct @abstractmethods:
https://stackoverflow.com/questions/46194493/determine-if-method-is-still-abstract

Reply:
For Player